src/lean_interact/interface.py

```diff
@@ -198,6 +200,248 @@ def message_intersects_code(msg: Message | Sorry, start_pos: Pos | None, end_pos
     return res
 
 
+class Range(BaseModel):
+    """Range of a Syntax object.
+    Attributes:
+        start: The starting position of the syntax.
+        finish: The ending position of the syntax.
+        synthetic: Boolean whether the syntax is synthetic or not.
+    """
+
+    synthetic: bool
+    start: Pos
+    finish: Pos
+
+    def __eq__(self, other):
+        return self.start == other.start and self.finish == other.finish
+
+
+class Syntax(BaseModel):
+    """Lean Syntax object.
+    Attributes:
+        pp: Pretty-printed string of the syntax.
+        range: Range of the syntax.
+        kind: SyntaxNodeKind for the syntax.
+        arg_kinds: SyntaxNodeKinds for the children of the syntax.
+    """
+
+    pp: str | None
+    range: Range
+    kind: str
+    arg_kinds: list[str] = Field(default_factory=list, alias="argKinds")
+
+
+class BaseNode(BaseModel):
+    """Base for the nodes of the InfoTree.
+    Attributes:
+        stx: Syntax object of the node.
+    """
+
+    stx: Syntax
+
+
+class TacticNode(BaseNode):
+    """A tactic node of the InfoTree.
+    Attributes:
+        stx: Syntax object of the node.
+        name: Optional name of the tactic.
+        goals_before: Goals before tactic application.
+        goals_after: Goals after tactic application.
+    """
+
+    name: str | None
+    goals_before: list[str] = Field(default_factory=list, alias="goalsBefore")
+    goals_after: list[str] = Field(default_factory=list, alias="goalsAfter")
+
+
+class CommandNode(BaseNode):
+    """A command node of the InfoTree.
+    Attributes:
+        stx: Syntax object of the node.
+        elaborator: The elaborator used to elaborate the command.
+    """
+
+    elaborator: str
+
+
+class TermNode(BaseNode):
+    """A term node of the InfoTree.
+    Attributes:
+        stx: Syntax object of the node.
+        is_binder: Whether the node is a binder or not.
+        expr: Expression string.
+        expected_type: Expected type.
+        elaborator: Optionally, the elaborator used.
+    """
+
+    is_binder: bool = Field(alias="isBinder")
+    expr: str
+    expected_type: str | None = Field(default=None, alias="expectedType")
+    elaborator: str | None
+
+
+Node = TacticNode | CommandNode | TermNode | None
```

Comment:
`self.node` can be `None` according to its typing annotation (in Lean, node is an option as well).
Yeah, otherwise, most of the checks I added are useless.
I guess something better to do would be to update the return type of the `commands` method to tell the user, and the type checker, it is a generator of `CommandNode`.

Reply:
Oh, but this is not trivial. It is returning `InfoTree`s with nodes that are `CommandNode`...

Reply:
Yeah I agree, let's merge like this